### PR TITLE
Add faster parsing + colbert embeddings

### PR DESCRIPTION
**Pull request by Dsantra92**
- vscode indexing how happens in less than 3 seconds compared to >47s run.
- Reorganized parsing to a better repo structure consistent with context graph release.

## Summary by CodeRabbit

* **New Features**
  * Python-first code-graph extraction API, optional Rust-backed parser, and optional hybrid Qdrant indexing (dense, sparse, multivector); Qdrant config support added.

* **Bug Fixes / Compatibility**
  * More tolerant handling of differing graph payload schemas during reconstruction.

* **Documentation**
  * Added parsing README with usage examples and semantics.

* **Infrastructure**
  * Updated packaging/manifests, Docker build contexts, CI tool version, expanded .gitignore, and added Qdrant env entries.

* **Tests**
  * Added unit tests for Python graph creation.

### CODE REVIEW

File: app/src/parsing/parsing/py_graph.py [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+            rel_fname=rel_fname,
+            fname=fname,
+            name=node_text,
+            kind=kind,
+            line=node.start_point[0],
+            end_line=node.end_point[0],
+            type=type,
+            byte_start=node.start_byte,
+            byte_end=node.end_byte,
+        )
```
[the remark this answers is not in the document]
Comment (akothari21) [reply app/src/parsing/parsing/py_graph.py:451]: @Dsantra92 fix this? also add a test case that replicates this behavior, as the bot comment suggests.

File: app/src/parsing/src/tag_extract.rs [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+                let Some(source_type) = node_types.get(&reference.source_id) else {
+                    continue;
+                };
+                let Some(target_type) = node_types.get(&target_metadata.node_id) else {
+                    continue;
+                };
+
+                if !is_valid_reference_direction(&reference.source_id, source_type, target_type) {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (akothari21) [reply app/src/parsing/src/tag_extract.rs:161]: @Dsantra92 any rationale for preventing reference edges from being created? Our existing test ( `test_extract_graph`)  is a single file check, so these edge cases never get tested. Can we add multi-directory tests too?

File: app/modules/parsing/knowledge_graph/qdrant_indexing_service.py [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+            logger.info("ColBERT model loaded via transformers")
+        else:
+            logger.info(
+                "Loading ColBERT model via pylate models.ColBERT: %s",
+                COLBERT_MODEL,
+            )
+            _colbert_model = pylate_models.ColBERT(model_name_or_path=COLBERT_MODEL)
+            _colbert_tokenizer = getattr(_colbert_model, "tokenizer", None)
+            logger.info("ColBERT model loaded via pylate")
+    return _colbert_model, _colbert_tokenizer
```
[the remark this answers is not in the document]
Comment (akothari21) [reply app/modules/parsing/knowledge_graph/qdrant_indexing_service.py:454]: we'll get a shape error ( ValueError ) when enforcing  `arr.shape[1] == COLBERT_DIM` using manual tokenization if `pylate` is not installed.

File: app/src/parsing/src/tag_extract.rs [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+                    relationship_type: "CONTAINS".to_string(),
+                    ident: Some(tag.name.clone()),
+                    ref_line: None,
+                    end_ref_line: None,
+                });
+            }
+
+            definitions.push((tag.name.clone(), node_id));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (akothari21) [reply app/src/parsing/src/tag_extract.rs]: @Dsantra92 another issue not being caught because of `test_extract_graph` not being a good enough test. Perhaps we should invest some time fixing the test / adding more tests to add different scenarios that the bot highlighted ( Classes, functions spread across different directories or multiple files, one non-FILE node etc )

**Review by yashkrishan — APPROVED** [review]
